File: Pharmaceutical_Company_Example/Behavioral/Python/Visitor.py
from abc import ABC, abstractmethod

# Define Element Interface for Regulatory Steps
class RegulatoryStep(ABC):
    @abstractmethod
    def accept(self, visitor):
        pass

# Concrete Elements
class DocumentationSubmission(RegulatoryStep):
    def accept(self, visitor):
        visitor.visit_documentation_submission(self)

class ClinicalTrialReview(RegulatoryStep):
    def accept(self, visitor):
        visitor.visit_clinical_trial_review(self)

class SafetyEvaluation(RegulatoryStep):
    def accept(self, visitor):
        visitor.visit_safety_evaluation(self)

# Visitor Interface
class Visitor(ABC):
    @abstractmethod
    def visit_documentation_submission(self, documentation):
        pass

    @abstractmethod
    def visit_clinical_trial_review(self, trial_review):
        pass

    @abstractmethod
    def visit_safety_evaluation(self, safety_evaluation):
        pass

# Concrete Visitor Auditor
class Auditor(Visitor):
    def visit_documentation_submission(self, documentation):
        print("Auditing documentation submission...")

    def visit_clinical_trial_review(self, trial_review):
        print("Auditing clinical trial review process...")

    def visit_safety_evaluation(self, safety_evaluation):
        print("Auditing safety evaluation...")

# Concrete Visitor Report Generator
class ReportGenerator(Visitor):
    def visit_documentation_submission(self, documentation):
        print("Generating report for documentation submission...")

    def visit_clinical_trial_review(self, trial_review):
        print("Generating report for clinical trial review...")

    def visit_safety_evaluation(self, safety_evaluation):
        print("Generating report for safety evaluation...")

# Concrete Visitor Approval Checker
class ApprovalChecker(Visitor):
    def visit_documentation_submission(self, documentation):
        print("Checking approval status of documentation submission...")

    def visit_clinical_trial_review(self, trial_review):
        print("Checking approval status of clinical trial review...")

    def visit_safety_evaluation(self, safety_evaluation):
        print("Checking approval status of safety evaluation...")

# Define the regulatory steps
documentation = DocumentationSubmission()
clinical_trials = ClinicalTrialReview()
safety_eval = SafetyEvaluation()

# Define visitors
auditor = Auditor()
report_generator = ReportGenerator()
approval_checker = ApprovalChecker()

# Use the Auditor visitor on different steps
documentation.accept(auditor)
clinical_trials.accept(auditor)
safety_eval.accept(auditor)

# Use the Report Generator visitor on different steps
documentation.accept(report_generator)
clinical_trials.accept(report_generator)
safety_eval.accept(report_generator)

# Use the Approval Checker visitor on different steps
documentation.accept(approval_checker)
clinical_trials.accept(approval_checker)
safety_eval.accept(approval_checker)

File: Pharmaceutical_Company_Example/Behavioral/Java/Visitor.java
// Define Element Interface for Regulatory Steps
interface RegulatoryStep {
    void accept(Visitor visitor);
}

// Concrete Elements
class DocumentationSubmission implements RegulatoryStep {
    public void accept(Visitor visitor) {
        visitor.visitDocumentationSubmission(this);
    }
}

class ClinicalTrialReview implements RegulatoryStep {
    public void accept(Visitor visitor) {
        visitor.visitClinicalTrialReview(this);
    }
}

class SafetyEvaluation implements RegulatoryStep {
    public void accept(Visitor visitor) {
        visitor.visitSafetyEvaluation(this);
    }
}

// Visitor Interface
interface Visitor {
    void visitDocumentationSubmission(DocumentationSubmission documentation);
    void visitClinicalTrialReview(ClinicalTrialReview trialReview);
    void visitSafetyEvaluation(SafetyEvaluation safetyEvaluation);
}

// Concrete Visitor: Auditor
class Auditor implements Visitor {
    public void visitDocumentationSubmission(DocumentationSubmission documentation) {
        System.out.println("Auditing documentation submission...");
    }

    public void visitClinicalTrialReview(ClinicalTrialReview trialReview) {
        System.out.println("Auditing clinical trial review process...");
    }

    public void visitSafetyEvaluation(SafetyEvaluation safetyEvaluation) {
        System.out.println("Auditing safety evaluation...");
    }
}

// Concrete Visitor: Report Generator
class ReportGenerator implements Visitor {
    public void visitDocumentationSubmission(DocumentationSubmission documentation) {
        System.out.println("Generating report for documentation submission...");
    }

    public void visitClinicalTrialReview(ClinicalTrialReview trialReview) {
        System.out.println("Generating report for clinical trial review...");
    }

    public void visitSafetyEvaluation(SafetyEvaluation safetyEvaluation) {
        System.out.println("Generating report for safety evaluation...");
    }
}

// Concrete Visitor: Approval Checker
class ApprovalChecker implements Visitor {
    public void visitDocumentationSubmission(DocumentationSubmission documentation) {
        System.out.println("Checking approval status of documentation submission...");
    }

    public void visitClinicalTrialReview(ClinicalTrialReview trialReview) {
        System.out.println("Checking approval status of clinical trial review...");
    }

    public void visitSafetyEvaluation(SafetyEvaluation safetyEvaluation) {
        System.out.println("Checking approval status of safety evaluation...");
    }
}

// Main class to execute the Visitor pattern
public class Main {
    public static void main(String[] args) {
        // Define regulatory steps
        RegulatoryStep documentation = new DocumentationSubmission();
        RegulatoryStep clinicalTrials = new ClinicalTrialReview();
        RegulatoryStep safetyEval = new SafetyEvaluation();

        // Define visitors
        Visitor auditor = new Auditor();
        Visitor reportGenerator = new ReportGenerator();
        Visitor approvalChecker = new ApprovalChecker();

        // Apply the Auditor visitor to different steps
        documentation.accept(auditor);
        clinicalTrials.accept(auditor);
        safetyEval.accept(auditor);

        System.out.println();

        // Apply the Report Generator visitor to different steps
        documentation.accept(reportGenerator);
        clinicalTrials.accept(reportGenerator);
        safetyEval.accept(reportGenerator);

        System.out.println();

        // Apply the Approval Checker visitor to different steps
        documentation.accept(approvalChecker);
        clinicalTrials.accept(approvalChecker);
        safetyEval.accept(approvalChecker);
    }
}
